File: Recursion/subsequence_pattern/Subsequence2.py
def _helper(index: int, numbers: list[int], result: list[list[int]], current: list[int]) -> None:
    result.append(current.copy())
    for next_index in range(index, len(numbers)):
        if next_index > index and numbers[next_index - 1] == numbers[next_index]:
            continue
        current.append(numbers[next_index])
        _helper(next_index + 1, numbers, result, current)
        current.pop()


def subsets_with_dup(numbers: list[int]) -> list[list[int]]:
    numbers.sort()
    result = []
    _helper(0, numbers, result, [])
    return result

File: Recursion/subsequence_pattern/subsequence2.cpp
#include <algorithm>
#include <vector>

class subsequence2 {
    static void helper(int index, const std::vector<int>& numbers, std::vector<std::vector<int>>& result,
                       std::vector<int>& current) {
        result.push_back(current);
        for (int i = index; i < static_cast<int>(numbers.size()); ++i) {
            if (i > index && numbers[i - 1] == numbers[i]) continue;
            current.push_back(numbers[i]);
            helper(i + 1, numbers, result, current);
            current.pop_back();
        }
    }
public:
    static std::vector<std::vector<int>> subsetsWithDup(std::vector<int> numbers) {
        std::sort(numbers.begin(), numbers.end());
        std::vector<std::vector<int>> result;
        std::vector<int> current;
        helper(0, numbers, result, current);
        return result;
    }
};